### PR TITLE
Spike: Pause compaction when out of sync

### PR DESCRIPTION
**Pull request by ZenGround0**
## Related Issues

I ended up playing with some of the chain sync contention protection ideas from #10388 

## Proposed Changes

Use head change out of sync detection to signal compaction to pause.  Signal continue when chain is back in sync

## Additional Info

@vyzo if you can take a look at this for feasibility it would be a big help.

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

**Review by vyzo — COMMENTED** [review]
it is not unreasonable i think, but we should test it in mainnet node.

File: blockstore/badger/blockstore.go [hunk 1]
```diff
@@ -396,6 +396,9 @@ func (b *Blockstore) doCopy(from, to *badger.DB) error {
 	if workers < 2 {
 		workers = 2
 	}
+	if workers > 8 {
+		workers = 8
```
Comment (magik6k) [inline blockstore/badger/blockstore.go:400]: Do we have confidence in this number? Did we check that it's "good enough" on mainnet?

File: blockstore/badger/blockstore.go [hunk 2]
```diff
@@ -458,11 +461,15 @@ func (b *Blockstore) onlineGC(ctx context.Context, threshold float64) error {
...
 	}
-
+	checkTick := time.NewTimer(checkFreq)
+	defer checkTick.Stop()
 	for err == nil {
 		select {
 		case <-ctx.Done():
 			err = ctx.Err()
+		case <-checkTick.C:
+			check()
```
Comment (magik6k) [inline blockstore/badger/blockstore.go:471]: Ignoring error?

File: blockstore/splitstore/splitstore_compact.go [hunk 3]
```diff
@@ -90,7 +90,35 @@ func (s *SplitStore) HeadChange(_, apply []*types.TipSet) error {
...
+			if atomic.CompareAndSwapInt32(&s.outOfSync, 0, 1) {
+				// transition from in sync to out of sync
+				s.chainSyncMx.Lock()
+				s.chainSyncFinished = false
+				s.chainSyncMx.Unlock()
+			}
+			// already out of sync, no signaling necessary
+
+		}
+		// TODO: ok to use hysteresis with no transitions between 30s and 1m?
```
Comment (magik6k) [inline blockstore/splitstore/splitstore_compact.go:107]: When we're syncing, the head will only be taken when we get fully in sync, so this is probably fine

(well, really the sync new head is only selected when we complete the previous sync, but the effect is similar)

**Review by magik6k — COMMENTED** [review]
Don't see anything massively wrong here, definitely would be good to test on a mainnet node.

File: blockstore/badger/blockstore.go (hunk 1 above)
Comment (vyzo) [reply blockstore/badger/blockstore.go:400]: maybe 5-10 min.

Comment (vyzo) [reply blockstore/badger/blockstore.go:400]: lets take a fresh measurement?

Comment (ZenGround0) [reply blockstore/badger/blockstore.go:400]: For reference I got 50m on my fairly resourced mainnet node

Comment (ZenGround0) [reply blockstore/badger/blockstore.go:400]: ^ with current default of half CPUs
I'll measure with 8 and compare when I get to similar garbage levels

Comment (ZenGround0) [reply blockstore/badger/blockstore.go:400]: working without major issues on mainnet (cc @TippyFlitsUK ) and this seems to be preventing out of sync on moving GC

File: blockstore/splitstore/splitstore_compact.go [hunk 4]
```diff
@@ -884,6 +912,25 @@ func (s *SplitStore) waitForSync() {
 	}
 }
 
+// Block compaction operations if chain sync has fallen behind
+func (s *SplitStore) waitForSync() {
+	if atomic.LoadInt32(&s.outOfSync) == 0 {
+		return
+	}
+	s.chainSyncMx.RLock()
```
Comment (magik6k) [inline blockstore/splitstore/splitstore_compact.go:920]: Don't you want .Lock for the cond?

**Comment by Stebalien** [comment]
@arajasek I don't _think_ you meant to close this, right?

**Comment by ZenGround0** [comment]
Closed by #10641 backported and merged back to master

**Pull request closed without merging**